File: 2016/go/day10.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strconv"
	"strings"
)

func handle(err error) {
	if err != nil {
		fmt.Println(err)
		os.Exit(1)
	}
}

type BotFuncs interface {
	numChips()  // how many chips
	getChips()  // what chips
	loseChips() // gives chips away, loses them
	gainChip()  // receives chips
}

type Bot struct {
	chips    [2]int       // chips that the bot has
	commands [2][2]string // which bots/bins the bot gives its chips to (low then high)
}

func (x Bot) numChips() int {
	count := 0
	for _, chip := range x.chips {
		if chip != 0 {
			count++
		}
	}
	return count
}

func (x Bot) getChips() (int, int) {
	return x.chips[0], x.chips[1]
}

func (x Bot) loseChips() Bot {
	x.chips = [2]int{0, 0}
	return x
}

func (x Bot) gainChip(chip int) Bot {
	for i := 0; i < 2; i++ {
		if x.chips[i] == 0 {
			x.chips[i] = chip
			break
		}
	}
	return x
}

func (x Bot) loadCmds(cmds [2][2]string) Bot {
	x.commands = cmds
	return x
}

func (x Bot) getCmds() [2][2]string {
	return x.commands
}

func min(x, y int) int {
	if x < y {
		return x
	}
	return y
}

func max(x, y int) int {
	if x > y {
		return x
	}
	return y
}

func chipsInBins(output [21][]int) int {
	total := 0
	for i := 0; i < len(output); i++ {
		total += len(output[i])
	}
	return total
}

func distribute(bots [210]Bot, chip_total int, targetA int, targetB int) (int, int) {
	var output [21][]int // got size by: grep -o "output \d*" puzzle_input.txt | sort -u
	var targetBot int

	for chipsInBins(output) < chip_total {
		for i := 0; i < len(bots); i++ {
			if bots[i].numChips() == 2 {
				chipA, chipB := bots[i].getChips()
				if (chipA == targetA && chipB == targetB) || (chipA == targetB && chipB == targetA) {
					targetBot = i
				}
				commands := bots[i].getCmds()

				for j := 0; j < 2; j++ {
					var chip int
					if j == 0 {
						chip = min(chipA, chipB)
					} else {
						chip = max(chipA, chipB)
					}
					num, err := strconv.Atoi(string(commands[j][1]))
					handle(err)
					if string(commands[j][0]) != "output" {
						bots[num] = bots[num].gainChip(chip)
					} else {
						output[num] = append(output[num], chip)
					}
				}
				bots[i] = bots[i].loseChips()
			}
		}
	}
	return targetBot, (output[0][0] * output[1][0] * output[2][0])
}

func main() {
	path := "puzzle_input.txt"
	file, err := ioutil.ReadFile(path)
	handle(err)
	lines := strings.Split(string(file), "\n")

	var bots [210]Bot // got size by: grep -o "bot \d\d\d" puzzle_input.txt | sort -u

	chip_total := 0
	for _, line := range lines {
		if line != "" {
			input := strings.Split(line, " ")
			if input[0] == "value" {
				chip, err := strconv.Atoi(input[1])
				handle(err)
				chip_total++
				bot, err := strconv.Atoi(input[5])
				handle(err)
				bots[bot] = bots[bot].gainChip(chip)
			} else {
				bot, err := strconv.Atoi(input[1])
				handle(err)
				commands := [2][2]string{
					{input[5], input[6]},
					{input[10], input[11]},
				}
				bots[bot] = bots[bot].loadCmds(commands)
			}
		}
	}

	part1, part2 := distribute(bots, chip_total, 61, 17)
	fmt.Printf("Part 1: Bot #%d\n", part1)
	fmt.Println("Part 2:", part2)
}

File: 2016/Day_10/day10.java
import java.io.File;
import java.util.Scanner;
import java.util.ArrayList;

public class day10 {
    public static void main(String[] args) 
    {
        String path = "puzzle_input.txt";
        File file = new File(path);
        String[] lines = new String[231];
        try {
            Scanner scanner = new Scanner(file);
            int index = 0;
            while (scanner.hasNextLine()) {
                lines[index++] = scanner.nextLine();
            }
            scanner.close();
        } catch (Exception e) {
            e.printStackTrace();
            System.exit(0);
        }
        
        Bot[] bots = new Bot[210]; // got size by: grep -o "bot \d\d\d" puzzle_input.txt | sort -u
        for (int i = 0; i < bots.length; i++) {
            bots[i] = new Bot();
        }

        int chipTotal = 0;
        for (String line: lines) {
            String[] input = line.split(" ");
            if (input[0].equals("value")) {
                int chip = Integer.parseInt(input[1]);
                chipTotal++;
                int bot = Integer.parseInt(input[5]);
                bots[bot].gainChip(chip);
            } else {
                int bot = Integer.parseInt(input[1]);
                String[][] commands = new String[][]{
                    {input[5], input[6]},
                    {input[10], input[11]}
                };
                bots[bot].loadCmds(commands);
            }
        }
        int[] results = distribute(bots, chipTotal, 61, 17);
        System.out.println("Part 1: Bot #" + results[0]);
        System.out.println("Part 2: " + results[1]);
    }

    static int min(int x, int y)
    {
        if (x < y) {
            return x;
        }
        return y;
    }

    static int max(int x, int y)
    {
        if (x > y) {
            return x;
        }
        return y;
    }

    static int chipsInBins(ArrayList<ArrayList<Integer>> output)
    {
        int total = 0;
        for (int i = 0; i < output.size(); i++) {
            total += output.get(i).size();
        }
        return total;
    }

    static int[] distribute(Bot[] bots, int chipTotal, int targetA, int targetB) 
    {
        int size = 21; // got size by: grep -o "output \d*" puzzle_input.txt | sort -u
        ArrayList<ArrayList<Integer>> output = new ArrayList<ArrayList<Integer>>(size);
        for (int i = 0; i < size; i++) {
            output.add(new ArrayList<Integer>());
        }
        int targetBot = 0;

        while (chipsInBins(output) < chipTotal) {
            for (int i = 0; i < bots.length; i++) {
                if (bots[i].numChips() == 2) {
                    int[] chips = bots[i].getChips();
                    if ((chips[0] == targetA && chips[1] == targetB) || (chips[1] == targetA && chips[0] == targetB)) {
                        targetBot = i;
                    }
                    String[][] commands = bots[i].getCmds();           
                    for (int j = 0; j < 2; j++) {
                        int chip;
                        if (j == 0) {
                            chip = min(chips[0], chips[1]);
                        } else {
                            chip = max(chips[0], chips[1]);
                        }
                        int num = Integer.parseInt(commands[j][1]);
                        if (!commands[j][0].equals("output")) {
                            bots[num].gainChip(chip);
                        } else {
                            output.get(num).add(chip);
                        }
                    }
                    bots[i].loseChips();
                }
            }
        }
        int product = 1;
        for (int i = 0; i < 3; i++) {
            product *= output.get(i).get(0);
        }
        return new int[] { targetBot, product };
    }
}
